xmlrpc/man/tcosxmlrpc.1
.TH tcosxmlrpc 1 "Jun 19, 2006" "tcosxmlrpc man page"

.SH NAME
tcosxmlrpc \- TCOS XML-RPC server

.SH SYNOPSIS
.B tcosxmlrpc
.RI [ options ]
.br

.SH DESCRIPTION

This manpage show options for 
.B tcosxmlrpc
server.

.PP
\fBtcosxmlrpc\fP is a small xmlrpc server that listen xmlrpc connections
from TcosMonitor and allow to exec, or scan status of thin clients.

.SH OPTIONS
.B
.IP /path/to/abyss.conf
Path (relative or absolute) to a well formated abyss.conf file

.B
.IP TCOS_DEBUG
Set TCOS_DEBUG=1 to see debug output

.SH BUGS
Contact Mario Izquierdo <mariodebian@gmail.com>

Homepage: http://www.tcosproject.org

.SH SEE ALSO
tcos.conf(5)

.SH AUTHOR
Mario Izquierdo (mariodebian) <mariodebian@gmail.com>


debian/changelog
initramfs-tools-tcos (0.84.5) unstable; urgency=low

  * Remove tcos-server-utils package (now contents are in tcos package)
  * Remove obsolete opengl dir
  * Remove obsolete discover dir
  * Remove obsolete esound dir
  * Remove old dir

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Mon, 24 Sep 2007 13:10:18 +0200

initramfs-tools-tcos (0.84.4) unstable; urgency=low

  * Make TCOS_EXTRA_MODULES new var in tcos.conf
  * gentcos:
    - Remove export TCOS_EXTRA_MODULES
    - Fix read -extramodules cmdline to allow more modules

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Mon, 24 Sep 2007 13:07:23 +0200

initramfs-tools-tcos (0.84.3) unstable; urgency=low

  * Only depend on mknbi on i386 arch
  * Patch debian/control in Ubuntu dapper with Source-Version

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Sat, 22 Sep 2007 21:23:44 +0200

initramfs-tools-tcos (0.84.2) unstable; urgency=low

  * bin/tcos-buildroot:
    - Fix typo in TCOS_CHROOT var
    - Add --gentcos option to build images with other params
  * Update Ubuntu gutsy kernel version
  * debian/control:
    - Replace source:version depends with 0.84.1
    - Fix some descriptions 

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Sat, 22 Sep 2007 18:52:55 +0200

initramfs-tools-tcos (0.84.1) unstable; urgency=low

  * Important move of code, some of tcosmonitor stuff move here:
    - udev
    - xmlrpc
    - lockscreen
    - hooks-addons
  * bin/tcos-buidchroot:
    - Script to build a (32bit) chroot to build tcos images

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Fri, 21 Sep 2007 23:32:14 +0200

initramfs-tools-tcos (0.83.2) unstable; urgency=low

  * bin/configurexorg, hooks-addons/40etc3800:
    - Support new openchrome driver
  * scripts/tcos-top/20mountnfs:
    - Mount NFS without nfsro (deprecated in new unionfs)
  * scripts/tcos-top/50fstab:
    - Call udevtrigger and udevsettle to create ide/block devices
  * scripts/tcos-premount/25nbd_filesystem:
    - Create only 2 nbd devices and wait for /dev/nbd0

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Wed, 19 Sep 2007 22:41:27 +0200

initramfs-tools-tcos (0.83.1) unstable; urgency=low

  * scripts/tcos-bottom/05inittcos, conf/tcos.conf
    - limit MAX_MEM used for thin clients with ulimit (default 95%)
  * Remove PATH definition in all scripts
  * conf/tcos-run-functions.sh:
    - Disable change vt in kill_usplash
  * scripts/tcos-top/50fstab:
    - Add option to force not use swap
    - Replace cut with awk
  * scripts/tcos-bottom/10ldconfig, hooks/tcosbins:
    - Move ldconfig to /usr/sbin
  * bin/gentcos:
    - Fix typo in inst_pkg function
  * hooks-addons/08discover:
    - Move discover /lib files to squashed /usr/lib (less ram)

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Fri, 14 Sep 2007 21:32:49 +0200

initramfs-tools-tcos (0.82.11) unstable; urgency=low

  * Add USB/IP experimental support

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Wed, 05 Sep 2007 23:06:21 +0200

initramfs-tools-tcos (0.82.10) unstable; urgency=low

  * Update debian unstable kernel version 2.6.22-2-486

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Wed, 05 Sep 2007 22:04:44 +0200

initramfs-tools-tcos (0.82.9) unstable; urgency=low

  * bin/gentcos, hooks-addons/50xorg:
    - Add -fix-nvidia-ati option to cache xserver-xorg-core and libgl1-mesa-glx
    - Reorganize bin/gentcos with simpler code
  * bin/tcos-gdm-autologin:
    - Remove wrong licence line (thanks to Quiran to alert me)
  * daemonize/daemon.c, daemonize/Makefile:
    - Make work daemonize with double fork.

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Tue, 04 Sep 2007 11:10:17 +0200

initramfs-tools-tcos (0.82.8) unstable; urgency=low

  * Add support for cdaudio (using cdfs + cdparanoia to detect)
  * Update dapper kernel 2.6.15-29-386

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Sat, 01 Sep 2007 13:18:30 +0200

initramfs-tools-tcos (0.82.7) unstable; urgency=low

  * bin/get_filesystem:
    - Remove disktype
    - Add fstype (new klibc)
  * Update Ubuntu gutsy kernel version

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Fri, 24 Aug 2007 17:59:18 +0200

initramfs-tools-tcos (0.82.6) unstable; urgency=low

  * bin/tcos-gdm-autologin:
    - New option (--noaction) for debug proposes only
    - If DISPLAY contains IP address instead of hostname try to convert using gethostbyaddr
    - If /etc/gd/gdm.conf-custom exists use it.
    - Update man page
  * conf/tcos-run-functions.sh:
    - Kill tryXorg too in kill_xorg()
    - Kill process in background

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Tue, 21 Aug 2007 23:10:18 +0200

initramfs-tools-tcos (0.82.5) unstable; urgency=low

  * Add Subversion repository URL to debian/control.
  * bin/gentcos:
    - Ignore errors in ramdisk size

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Mon, 20 Aug 2007 18:28:27 +0200

initramfs-tools-tcos (0.82.4) unstable; urgency=low

  * Add support for lp_server (if avalaible)
  * Change Debian etch kernel version to 2.6.18-5-486

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Mon, 20 Aug 2007 11:08:59 +0200

initramfs-tools-tcos (0.82.3) unstable; urgency=low

  * Update kernel versions (enable update && security mirrors in pbuilder)

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Tue, 07 Aug 2007 13:32:34 +0200

initramfs-tools-tcos (0.82.2) unstable; urgency=low

  * Fix new gentcos test about unionfs|aufs module (crash without awk)
  * Create /sbin/startesd always not only with cached esound

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Mon, 30 Jul 2007 23:39:09 +0200

initramfs-tools-tcos (0.82.1) unstable; urgency=low

  * Add printer support TCOS_PRINT
  * Add initial NBD support (don't work fully yet)
  * Add support for using aufs module instead of unionfs

xmlrpc/tcosxmlrpc.c
/*# tcosxmlrpc.c standalone XML-RPC server written in C 2006-09-09 14:22:40 mariodebian $
#
# This file is part of tcosxmlrpc.
#
# tcosxmlrpc is free software; you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# tcosxmlrpc is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with tcosxmlrpc; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
# USA.
*/

/* 

Incompatible pointer types error
http://www.elrincondelc.com/portal/modules.php?name=Forums&file=viewtopic&p=20325

*/



#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <xmlrpc.h>

#ifdef XMLRPC_INC
  #include XMLRPC_INC
#endif

#include <unistd.h>

/*#define VERSION "0.0.15"*/
#define BUFF_SIZE 1000
#define HAVE_MAIN

/* My TCOS includes */

#include "debug.c"

#include "login.c"
#include "xauth.c"
#include "info.c"
#include "standalone.c"
#include "pci.c"
#include "exe.c"
#include "screenshot.c"
#include "xorg.c"
#include "sound.c"
#include "devices.c"
#include "lockscreen.c"

static xmlrpc_value *
tcos_version(xmlrpc_env *env, xmlrpc_value *in, void *ud)
 {

    dbgtcos("tcosxmlrpc::tcos_version() %s\n", VERSION);
    return xmlrpc_build_value(env, "s", VERSION);
 }

static xmlrpc_value *
tcos_echo(xmlrpc_env *env, xmlrpc_value *in, void *ud)
 {
   char *s;
   size_t *len;

   
   xmlrpc_parse_value(env, in, "(s#)", &s, &len);

   dbgtcos("tcosxmlrpc::tcos_echo() %s\n", s);

   return xmlrpc_build_value(env, "s", s);
 }



static xmlrpc_value *
tcos_status (xmlrpc_env *env, xmlrpc_value *param_array, void *user_data)
{
   FILE *fp=NULL;
   char *app=NULL;
   /*char *login_ok;
   char *user;
   char *pass;*/
   char cmd[BUFF_SIZE];
   char ret[BUFF_SIZE];


   dbgtcos("tcosxmlrpc::tcos_status() Init \n");

   /* Parse app string */
   xmlrpc_parse_value(env, param_array, "(s)", &app);

/*   xmlrpc_parse_value(env, param_array, "(sss)", &app, &user, &pass); */
   if (env->fault_occurred)
	return xmlrpc_build_value(env, "s", "params error");

   /* need login first */
  /* login_ok=validate_login(user,pass);
  if( strcmp(login_ok,  LOGIN_OK ) != 0 )
    return xmlrpc_build_value(env, "s", login_ok );
  */

   dbgtcos("tcosxmlrpc::tcos_status() pidof %s\n", app);
   

   snprintf( (char*) &cmd, BUFF_SIZE, "pidof %s| grep [1234567890] | wc -l" ,app);

   dbgtcos("tcosxmlrpc::tcos_status() exec cmd=\"%s\"\n", cmd);

   fp=(FILE*)popen(cmd, "r");
   if (env->fault_occurred)
	return xmlrpc_build_value(env, "s", "exec error");;

   dbgtcos("tcosxmlrpc::tcos_status() reading from fp pointer\n");

   fscanf(fp, "%s", ret);

   dbgtcos( "tcosxmlrpc::tcos_status() ret value=%s\n", ret);
   
   pclose(fp);

   if (ret != NULL)   
       return xmlrpc_build_value(env, "s", ret);
   else
       return xmlrpc_build_value(env, "s", "error");

}


int main (int argc, char **argv)
{
    /* check command line config file */
    if (argc != 2) {
	fprintf(stderr, "Usage:\n\t tcosxmlrpc /path/of/abyss.conf\n");
	return(1);
    }


    xmlrpc_server_abyss_init(XMLRPC_SERVER_ABYSS_NO_FLAGS, argv[1]);

    /* add methods */

    xmlrpc_server_abyss_add_method_w_doc("tcos.login", &tcos_login, NULL,
    "ss:s", "Tcos, Basic auth system.");

    xmlrpc_server_abyss_add_method_w_doc("tcos.logout", &tcos_logout, NULL,
    ":s", "Tcos, Basic auth system.");

    xmlrpc_server_abyss_add_method_w_doc("tcos.version", &tcos_version, NULL,
    ":s", "Tcos, Returns tcosxmlrpc version. (no auth needed)");

    xmlrpc_server_abyss_add_method_w_doc("tcos.exe", &tcos_exe, NULL,
    "sss:s", "Tcos, Exec a command passed as string. (need login first)");

    xmlrpc_server_abyss_add_method_w_doc("tcos.kill", &tcos_kill, NULL,
    "sss:s", "Tcos, killall command passed as string. (need login first)");

    xmlrpc_server_abyss_add_method_w_doc("tcos.status", &tcos_status, NULL,
    "s:s", "Tcos, returns 1 if app running 0 if not or error string. (no auth needed)");

    xmlrpc_server_abyss_add_method_w_doc("tcos.screenshot", &tcos_screenshot, NULL,
    "ss:s", "Tcos, make a screenshot and return files. (need login first)");

    xmlrpc_server_abyss_add_method_w_doc("tcos.xorg", &tcos_xorg, NULL,
    "ssss:s", "Tcos, configure, change or get xorg.conf settings. (need login first)");

    xmlrpc_server_abyss_add_method_w_doc("tcos.sound", &tcos_sound, NULL,
    "ssss:s", "Tcos, configure, change or get sound settings.");

    xmlrpc_server_abyss_add_method_w_doc("tcos.devices", &tcos_devices, NULL,
    "ssss:s", "Tcos, configure, change or get devices settings.");

    xmlrpc_server_abyss_add_method_w_doc("tcos.xauth", &tcos_xauth, NULL,
    "ss:s", "Tcos, authenticate with X cookies.");

    xmlrpc_server_abyss_add_method_w_doc("tcos.lockscreen", &tcos_lockscreen, NULL,
    "ss:s", "Tcos, exec lockscreen to block thin client.");

    xmlrpc_server_abyss_add_method_w_doc("tcos.unlockscreen", &tcos_unlockscreen, NULL,
    "ss:s", "Tcos, kill lockscreen to unblock thin client.");

/* DEPRECATED METHOD
    xmlrpc_server_abyss_add_method_w_doc("tcos.update_system_info", &tcos_update_system_info, NULL,
    ":i", "Exec \"WEBSERVER_SCRIPT update\" (need login first).");
*/
    xmlrpc_server_abyss_add_method_w_doc("tcos.echo", &tcos_echo, NULL,
    "s:s", "Tcos, debug class, returns string passed. (no auth needed)");

    /* Info method */
    xmlrpc_server_abyss_add_method_w_doc("tcos.info", &tcos_info, NULL,
    "s:s", "Tcos, return passed info string. (no auth needed)\n\
Info methods:\n\
* cpu_model\n\
* cpu_speed\n\
* cpu_vendor\n\
* ram_total\n\
* ram_active\n\
* ram_free\n\
* ram_used\n\
* swap_avalaible\n\
* swap_total\n\
* swap_used\n\
* swap_free\n\
* tcos_date\n\
* tcos_generation_date\n\
* tcos_version\n\
* network_hostname\n\
* network_ip\n\
* network_mak\n\
* network_mask\n\
* network_rx\n\
* network_tx\n\
* modules_loaded");

    xmlrpc_server_abyss_add_method_w_doc("tcos.standalone", &tcos_standalone, NULL,
    "s:s", "Tcos, Standalone. Return standalone values (no auth needed)");

    xmlrpc_server_abyss_add_method_w_doc("tcos.dbus", &tcos_dbus, NULL,
    "s:s", "Tcos, DBus. Exe with a wrapper some dbus events (auth needed)");

    xmlrpc_server_abyss_add_method_w_doc("tcos.pci", &tcos_pci, NULL,
    "s:s", "Tcos, PCI data stuff. Send pci_all to get pci bus ids. (no auth needed)");

    dbgtcos("tcosxmlrpc::main() switching to background.\n");

    xmlrpc_server_abyss_run();

    /* We never reach this point. */
    return 0;
}
